Support graceful Linux/Windows VM shutdown

jstachowiak opened the pull request:
Closes #8469

jackofallops commented:
hi @jstachowiak - just passing through and noticed you've included the changelog in this PR.  Maintainers update that when we merge PR's so it's accurate and current for the release that the change goes into, so we'll need you to remove that change please. 
Thanks!

jstachowiak commented:
```
golangci/golangci-lint info checking GitHub for tag 'v1.24.0'
golangci/golangci-lint crit unable to find 'v1.24.0' - use 'latest' or see https://github.com/golangci/golangci-lint/releases for details
GNUmakefile:14: recipe for target 'tools' failed
make: *** [tools] Error 1
The command "make tools" failed and exited with 2 during .
Your build has been stopped.
```
I can see the tag, so I assume it's a glitch. How can I rerun the jobs?

jstachowiak commented:
Cool @jackofallops! I removed that change and added a few acceptance tests. Some jobs failed with the following error:
```
golangci/golangci-lint info checking GitHub for tag 'v1.24.0'
golangci/golangci-lint crit unable to find 'v1.24.0' - use 'latest' or see https://github.com/golangci/golangci-lint/releases for details
GNUmakefile:14: recipe for target 'tools' failed
make: *** [tools] Error 1
The command "make tools" failed and exited with 2 during .
Your build has been stopped.
```
I can see the tag, so I assume it's a glitch. How can I rerun these jobs?

tombuildsstuff commented on azurerm/internal/services/compute/linux_virtual_machine_resource.go:
```diff
@@ -208,6 +208,12 @@ func resourceLinuxVirtualMachine() *schema.Resource {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
```
since this field controls the behaviour - this wants to go [into the features block](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/internal/provider/features.go#L18-L21) instead of on the resource - can we move this there? That should then be available as `meta.(*clients.Client).Features.VirtualMachine.GracefulShutdown`

jstachowiak replied:
This makes a lot of sense. I'm going to use `graceful_shutdown` as the argument name in the `virtual_machine` block.

tombuildsstuff commented on azurerm/internal/services/compute/tests/linux_virtual_machine_resource_other_test.go:
```diff
@@ -521,6 +521,46 @@ func TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabledWithCMK(t *testing.T
+					resource.TestCheckResourceAttr(data.ResourceName, "skip_shutdown", "true"),
+				),
+			},
+			data.ImportStep("skip_shutdown"),
```
by moving this to the features block this is no longer necessary:

```suggestion
			data.ImportStep(),
```

tombuildsstuff commented on azurerm/internal/services/compute/tests/linux_virtual_machine_resource_other_test.go:
```diff
@@ -521,6 +521,46 @@ func TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabledWithCMK(t *testing.T
+					resource.TestCheckResourceAttr(data.ResourceName, "skip_shutdown", "false"),
+				),
+			},
+			data.ImportStep("skip_shutdown"),
```
```suggestion
			data.ImportStep(),
```

tombuildsstuff commented on azurerm/internal/services/compute/tests/linux_virtual_machine_resource_other_test.go:
```diff
@@ -521,6 +521,46 @@ func TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabledWithCMK(t *testing.T
+				Config: testLinuxVirtualMachine_otherSkipShutdownDisabled(data),
+				Check: resource.ComposeTestCheckFunc(
+					checkLinuxVirtualMachineExists(data.ResourceName),
+					resource.TestCheckResourceAttr(data.ResourceName, "skip_shutdown", "false"),
```
we can remove this:

```suggestion
```

tombuildsstuff commented on azurerm/internal/services/compute/tests/linux_virtual_machine_resource_other_test.go:
```diff
@@ -521,6 +521,46 @@ func TestAccLinuxVirtualMachine_otherEncryptionAtHostEnabledWithCMK(t *testing.T
+				Config: testLinuxVirtualMachine_otherSkipShutdownDefault(data),
+				Check: resource.ComposeTestCheckFunc(
+					checkLinuxVirtualMachineExists(data.ResourceName),
+					resource.TestCheckResourceAttr(data.ResourceName, "skip_shutdown", "true"),
```
we can remove this:

```suggestion
```

tombuildsstuff commented on azurerm/internal/services/compute/tests/linux_virtual_machine_resource_other_test.go:
```diff
@@ -1573,3 +1613,75 @@ resource "azurerm_linux_virtual_machine" "test" {
+func testLinuxVirtualMachine_otherSkipShutdownDisabled(data acceptance.TestData) string {
+	template := testLinuxVirtualMachine_template(data)
+	return fmt.Sprintf(`
+%s
```
instead we can specify this feature in the `provider` block, which in practice should be on every test rather than defined in the template - since the conditions can vary per-test

it's worth noting this won't work right now (insofar as all tests will run with the same provider) - but that's something that's being worked on - so if we can add this in preparation we're testing that going forward 👍

jstachowiak replied:
Cool, I'm going to stop using the template and enable/disable the feature per-test.

tombuildsstuff commented on azurerm/internal/services/compute/tests/windows_virtual_machine_resource_other_test.go:
```diff
@@ -773,6 +773,52 @@ func TestAccWindowsVirtualMachine_otherEncryptionAtHostEnabledWithCMK(t *testing
+				Config: testWindowsVirtualMachine_otherSkipShutdownDefault(data),
+				Check: resource.ComposeTestCheckFunc(
+					checkWindowsVirtualMachineExists(data.ResourceName),
+					resource.TestCheckResourceAttr(data.ResourceName, "skip_shutdown", "true"),
```
(same here) we can remove this since this is instead set on the provider block:

```suggestion
```

tombuildsstuff commented on azurerm/internal/services/compute/tests/windows_virtual_machine_resource_other_test.go:
```diff
@@ -773,6 +773,52 @@ func TestAccWindowsVirtualMachine_otherEncryptionAtHostEnabledWithCMK(t *testing
+			},
+			data.ImportStep(
+				"admin_password",
+				"skip_shutdown",
```
(as above) we can remove this since this is set on the provider block instead:

```suggestion
```

tombuildsstuff commented on azurerm/internal/services/compute/tests/windows_virtual_machine_resource_other_test.go:
```diff
@@ -773,6 +773,52 @@ func TestAccWindowsVirtualMachine_otherEncryptionAtHostEnabledWithCMK(t *testing
+				Config: testWindowsVirtualMachine_otherSkipShutdownDisabled(data),
+				Check: resource.ComposeTestCheckFunc(
+					checkWindowsVirtualMachineExists(data.ResourceName),
+					resource.TestCheckResourceAttr(data.ResourceName, "skip_shutdown", "false"),
```
(same here)

```suggestion
```

tombuildsstuff commented on azurerm/internal/services/compute/tests/windows_virtual_machine_resource_other_test.go:
```diff
@@ -773,6 +773,52 @@ func TestAccWindowsVirtualMachine_otherEncryptionAtHostEnabledWithCMK(t *testing
+			},
+			data.ImportStep(
+				"admin_password",
+				"skip_shutdown",
```
(same here)

```suggestion
```

tombuildsstuff commented on azurerm/internal/services/compute/windows_virtual_machine_resource.go:
```diff
@@ -221,6 +221,12 @@ func resourceWindowsVirtualMachine() *schema.Resource {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
```
(same here) can we move this to the `features` block since this is a behavioural feature toggle?

tombuildsstuff reviewed (changes requested):
hey @jstachowiak 

Thanks for this PR :)

Taking a look through this is off to a good start - but since this is a behavioural feature toggle it'd be good to move this to the `features` block within the `provider` block to group the behavioural feature toggles together - could we update this?

Thanks!

jstachowiak commented:
Hi @tombuildsstuff! I implemented changes that you requested. `make depscheck` job failed but not sure why. Let me know if you have any other suggestions.

jstachowiak commented:
Hi @tombuildsstuff! I implemented the changes that you requested. `make depscheck` job failed but not sure why. Let me know if you have any other suggestions.

jstachowiak commented:
@tombuildsstuff, do you have time to take look at my changes?

jstachowiak commented:
Hi @tombuildsstuff, any update here? 🤔

tombuildsstuff commented:
@jstachowiak apologies for the delay, missed this - will take a look now 👍

tombuildsstuff commented on website/docs/index.html.markdown:
```diff
@@ -193,6 +193,10 @@ The `virtual_machine` block supports the following:
 
+* `graceful_shutdown` - (Optional) Should the `azurerm_linux_virtual_machine` and `azurerm_windows_virtual_machine` request a graceful shutdown when the Virtual Machine is destroyed? Defaults to `false`.
+
+~> **Note:** The Azure platform cannot wait forever after [initiating a graceful shutdown](https://azure.microsoft.com/en-us/blog/linux-and-graceful-shutdowns-2/). The platform will wait 5 minutes for the Virtual Machine to shut down gracefully, and if it is still running after that time it will power off the machine.
```
could we make this:

```suggestion
~> **Note:** When using a graceful shutdown, Azure gives the Virtual Machine a 5 minutes window in which to complete the shutdown process, at which point the machine will be force powered off - [more information can be found in this blog post](https://azure.microsoft.com/en-us/blog/linux-and-graceful-shutdowns-2/).
```

tombuildsstuff reviewed (changes requested):
hey @jstachowiak

Thanks for pushing those changes - apologies for the delayed re-review here!

I've taken a look through and left one comment inline around wording - but if we can fix that and rebase this PR on top of master then this should otherwise be good to merge 👍 

Thanks!

tombuildsstuff reviewed (approved):
LGTM - thanks for pushing those changes @jstachowiak

jstachowiak commented:
Thanks @tombuildsstuff - this was fun! 🚀

tombuildsstuff commented:
Acceptance Tests look good 👍
